### PR TITLE
[Relay] Add gradient operator tutorial docs

### PR DESCRIPTION
**Pull request by weberlo**
Another rough draft.  Roast me.

I'm interested in the answer to the TODO at the end of the patch.

@jroesch @MarisaKirisame @slyubomirsky @ZihengJiang

### CODE REVIEW

**Comment by ZihengJiang** [comment]
`collapse_sum_like` is only for operators which have broadcast behaviors

File: docs/dev/relay_add_op.rst [hunk 1]
```diff
@@ -139,6 +139,50 @@ before producing the call node:
         tup = Tuple(list(args))
         return _make.concat(tup)
 
+Gradient Operators
+------------------
+
+Gradient operators are important for writing differentiable programs in
```
Comment (slyubomirsky) [inline docs/dev/relay_add_op.rst]: Specifically you should say that gradient operators are necessary for the AD algorithm; the AD algorithm can differentiate first-class language constructs, but because operators are opaque, it needs an explicit differentiation rule for them.

File: docs/dev/relay_add_op.rst [hunk 2]
```diff
@@ -139,6 +139,50 @@ before producing the call node:
...
+
+The inputs here are the original operator and a gradient to accumulate into.
+What we return is a list, where the 0th index is the derivative of the
+multiply operator with respect to the first input, and the 1st index is the
+derivative with respect to the second input. In general, the gradient will
+return a list with as many elements as there are inputs to the base operator.
+
+Before we further analyze this definition, first we should recall the partial
+derivatives for multiplication. Given a function f(x, y) = x * y, we have
+that df/dx = y and df/dy = x. The definition above looks similar to the math
```
Comment (slyubomirsky) [inline docs/dev/relay_add_op.rst]: Do you know any way to use the proper partial derivative notation? ∂f/∂x?

File: docs/dev/relay_add_op.rst [hunk 3]
```diff
@@ -139,6 +139,50 @@ before producing the call node:
...
+definitions, but there are some subtle differences, which we describe below.
+
+We're not just interested in how to compute the gradient of this function.
+We're interested in composing this gradient with other gradients, so we can
+accumulate the gradient across an entire program. This is where the ``grad *
```
Comment (slyubomirsky) [inline docs/dev/relay_add_op.rst]: We may want a different choice of variable names because `grad * x` could be misinterpreted as ∇ * x (div). Also I think you should use :code: in front of the backticks

File: docs/dev/relay_add_op.rst (hunk 2 above)
Comment (weberlo) [reply docs/dev/relay_add_op.rst]: Wasn't sure if we wanted unicode in there.  I guess I could just add it and see if the docs break.

File: docs/dev/relay_add_op.rst (hunk 3 above)
Comment (weberlo) [reply docs/dev/relay_add_op.rst]: Are you sure?  I've looked at the source for some of the docs, and all they use are double backticks for code formatting.

Comment (weberlo) [reply docs/dev/relay_add_op.rst]: Also, I'm not sure what you mean when you say "∇ * x (div)".

File: docs/dev/relay_add_op.rst [hunk 4]
```diff
@@ -139,6 +139,55 @@ before producing the call node:
...
+y`` and ``grad * x`` terms come from. We know that ∂f/∂x = y, but the way we
+compose this derivative with the gradient thus far is by multiplying it.
+
+Additionally, since the shape of ``grad`` might not match the shape of either
+of the inputs, we use ``collapse_sum_like`` to take the contents of the
+``grad * <var>`` terms and make the shape match the input we're
+differentiating with respect to. We only need to do this for operators with
+broadcasting behaviors.
+
+TODO: Why do we only have ``collapse_sum_like`` on some of the gradient
```
Comment (MarisaKirisame) [inline docs/dev/relay_add_op.rst]: collapse is only needed when you broadcast, which most binary operation implicitly do

**Comment by weberlo** [comment]
Adding more reviewers!

@junrushao1994 @antinucleon

File: docs/dev/relay_add_op.rst (hunk 3 above)
Comment (slyubomirsky) [reply docs/dev/relay_add_op.rst]: https://en.wikipedia.org/wiki/Divergence

I don't know if other docs use something different (some kind of styling) but I have stuck to using `:code:` for Relay docs and found when I built docs that backticks without it ended up italicized, not with typewriter font.

**Comment by weberlo** [comment]
Ping @jroesch @MarisaKirisame @slyubomirsky @ZihengJiang @junrushao1994 @antinucleon

**Comment by slyubomirsky** [comment]
Could you try building the docs to make sure the inline ∂ symbols work? Also to make sure that the statements in backticks show up correctly in typewriter font. I haven't tried myself but I'd be concerned about those.

File: docs/dev/relay_add_op.rst [hunk 5]
```diff
@@ -139,6 +139,52 @@ before producing the call node:
...
+------------------
+
+Gradient operators are important for writing differentiable programs in
+Relay. While it is the case that Relay's autodiff algorithm can differentiate
+first-class language constructs, operators are opaque. Because Relay can't
+look into the implementation, an explicit differentiation rule must be
+provided.
+
+Adding a gradient operator is slightly different from adding a normal
+operator, in that you only need to touch Python code. A collection of
```
Comment (slyubomirsky) [inline docs/dev/relay_add_op.rst]: It can be done in C++ too (but generally has not been done this way). Is there any example of a gradient registered in C++? As far as I know, it can be done using the operator registry APIs

File: docs/dev/relay_add_op.rst [hunk 6]
```diff
@@ -139,6 +139,52 @@ before producing the call node:
...
+We're interested in composing this gradient with other gradients, so we can
+accumulate the gradient across an entire program. This is where the ``grad *
+y`` and ``grad * x`` terms come from. We know that ∂f/∂x = y, but the way we
+compose this derivative with the gradient thus far is by multiplying it.
+
+Additionally, since the shape of ``grad`` might not match the shape of either
+of the inputs, we use ``collapse_sum_like`` to take the contents of the
+``grad * <var>`` terms and make the shape match the input we're
+differentiating with respect to. We only need to do this for operators with
+broadcasting behaviors.
```
Comment (slyubomirsky) [inline docs/dev/relay_add_op.rst]: I think you should have another example that does not have the complexity of `collapse_sum_like` because this seems like a fairly confusing point (at least, I'm left confused). It would also benefit maybe from some lower-level explanation (or having some math more explicitly written out).

File: docs/dev/relay_add_op.rst (hunk 1 above)
Comment (slyubomirsky) [reply docs/dev/relay_add_op.rst]: Nits:
* Do we have any documentation of Relay's AD? It would be good to link to that
* It would be good to say that the differentiation rule is opaque to Relay as well
* I think the sentence in which you describe operators as "opaque" should also contain the clause clarifying that "opaque" means that Relay cannot look into the implementation

**Comment by weberlo** [comment]
> Could you try building the docs to make sure the inline ∂ symbols work? Also to make sure that the statements in backticks show up correctly in typewriter font. I haven't tried myself but I'd be concerned about those.

It seems that building the docs (assuming I'm doing it correctly) requires a CUDA-enabled GPU for some reason, which my machine doesn't have.  Not sure what to do about that.

File: docs/dev/relay_add_op.rst (hunk 6 above)
Comment (weberlo) [reply docs/dev/relay_add_op.rst]: Good point.  As far as I understand, it's to handle cases where you have a tensor `x` with shape `(4, 4)` and a tensor `y` with shape `(16, 1)`, and you add them.  When you differentiate wrt. `x`, you want the shape to match `x`, and you want the same for `y`.  I could be totally wrong though.

If one of the other reviewers can confirm that that's what's going on, then I can add that explanation to the docs.  Otherwise, adding another example is a good idea too.

File: docs/dev/relay_add_op.rst (hunk 5 above)
Comment (weberlo) [reply docs/dev/relay_add_op.rst]: With a quick grep, I couldn't find any instances of the "FPrimalGradient" attr being set in C++, so I suspect there aren't any C++ gradient examples.  I could still mention the procedure for adding one in C++ though.

**Comment by slyubomirsky** [comment]
Are you sure the build completely fails without CUDA enabled? I think some of the docs actually run TVM code and use it to generate example output, but those that don't do that should build

**Comment by slyubomirsky** [comment]
Are you sure the build completely fails without CUDA enabled? I think some of the docs actually run TVM code and use it to generate example output, but those that don't do that should build

**Comment by slyubomirsky** [comment]
Are you sure the build completely fails without CUDA enabled? I think some of the docs actually run TVM code and use it to generate example output, but those that don't do that should build

File: docs/dev/relay_add_op.rst (hunk 5 above)
Comment (slyubomirsky) [reply docs/dev/relay_add_op.rst]: Yeah, that's what I meant, just mention it. I don't think there are presently examples, but mention that it can be done and show how in case that proves to be convenient for others.

**Comment by weberlo** [comment]
> Are you sure the build completely fails without CUDA enabled? I think some of the docs actually run TVM code and use it to generate example output, but those that don't do that should build. (Also I'm pretty sure there's a LaTeX-like math mode in RST that some of the docs use, so perhaps look around for those to see a better way to encode mathematical expressions.)

They seem to be failing, because the build directory for the docs is empty when it finishes.

**Comment by slyubomirsky** [comment]
Have you tried a more targeted build that avoids the parts that require Cuda?

**Comment by weberlo** [comment]
> Have you tried a more targeted build that avoids the parts that require Cuda?

Didn't know about the `TVM_TUTORIAL_EXEC_PATTERN` environment variable.  Thanks for the offline help, Steven.

It seems that the unicode for partial derivatives shows up correctly, and you don't need `:code:` for inline snippets.  See the image below:

rendered_docs

**Comment by weberlo** [comment]
I'll add another gradient example when I get a chance.  Please feel free to review the current state of the PR though.

File: src/relay/pass/pattern_util.h [hunk 7]
```diff
@@ -309,6 +309,11 @@ inline Expr OnesLike(Expr e) {
   return CallNode::make(op, {e});
 }
 
+inline Expr CollapseSumLike(Expr e) {
+  static const Op& op = Op::Get("collapse_sum_like");
+  return CallNode::make(op, {e});
+}
+
```
Comment (weberlo) [inline src/relay/pass/pattern_util.h]: Should this definition be in a separate PR?
